packages/@aws-cdk/core/lib/stack-synthesizers/types.ts

```diff
@@ -31,4 +32,9 @@ export interface IStackSynthesizer {
    * Synthesize the associated stack to the session
    */
   synthesize(session: ISynthesisSession): void;
+
+  /**
+   * Returns another context with additional information, known by the synthesizer
+   */
+  enrichContext(ctx: cxschema.MissingContext): cxschema.MissingContext;
```

Comment:
Ooohhh this is very specific. Can't we just move this into the details of synthesis?

(I know they're a bit messy right now, half-divided between `Stack` and `_shared.ts`, but feels like it should be possible...?)

Reply:
I couldn't find a clean way to do this. Given that this function needs to use the `lookupRoleArn` prop, it has to be implemented in `DefaultSynthesizer`. The `Stack` class depends on `IStackSynthesizer`, which forces this method to be exposed in the API.

But if you can spot a better way of organizing this, I'd be happy to change it.

Reply:
I was thinking something like:

```ts
interface ISynthesizer {
  public queryContext(scope: Construct, provider: string, props: Record<string, string>): ContextQueryResult;
}
```

And we would move a bit of the functionality from here:

https://github.com/aws/aws-cdk/blob/40d2ff964c97954c70d79a09d60fcb795ef16791/packages/%40aws-cdk/core/lib/context-provider.ts#L93-L103

Into there.

Then the default synthesizer can enhance with the lookup role when calling `stack.reportMissingContextKey()`.

What do you think?